File: android/app/src/main/java/com/tanay/echo/data/Clock.kt
package com.tanay.echo.data

/**
 * A clock that never returns the same value twice and only ever moves forward, so every write
 * gets a unique, strictly-increasing updatedAt. That uniqueness is what lets the sync push
 * watermark use a strict `>` without ever stranding a row written in the same millisecond as a
 * prior push. A verbatim port of src/main/store/clock.ts. `timeSource` is injectable for tests.
 */
class MonotonicClock(private val timeSource: () -> Long = { System.currentTimeMillis() }) {
    private var last = 0L

    @Synchronized
    fun now(): Long {
        last = maxOf(timeSource(), last + 1)
        return last
    }
}

File: src/main/store/clock.ts
/**
 * A clock that never returns the same value twice and only ever moves forward, so every
 * write gets a unique, strictly-increasing `updated_at`. That uniqueness is what lets the
 * sync push watermark use a strict `>` comparison without ever stranding a row written in
 * the same millisecond as a prior push. `timeSource` is injectable for tests; production
 * uses the wall clock.
 */
export function monotonicClock(timeSource: () => number = () => Date.now()): () => number {
  let last = 0
  return () => {
    last = Math.max(timeSource(), last + 1)
    return last
  }
}
